src/output-tx.c

```diff
@@ -138,35 +148,87 @@ static TmEcode OutputTxLog(ThreadVars *tv, Packet *p, void *thread_data, PacketQ
         if (!(AppLayerParserStateIssetFlag(f->alparser, APP_LAYER_PARSER_EOF)))
         {
             int tx_progress = AppLayerParserGetStateProgress(p->proto, alproto,
-                                                             tx, FlowGetDisruptionFlags(f, STREAM_TOSERVER));
-            if (tx_progress < tx_progress_done_value_ts) {
-                SCLogDebug("progress not far enough, not logging");
-                break;
+                tx, FlowGetDisruptionFlags(f, STREAM_TOSERVER));
+            if (tx_progress == tx_progress_done_value_ts) {
+                ts_ready = 1;
             }
 
             tx_progress = AppLayerParserGetStateProgress(p->proto, alproto,
-                                                         tx, FlowGetDisruptionFlags(f, STREAM_TOCLIENT));
-            if (tx_progress < tx_progress_done_value_tc) {
-                SCLogDebug("progress not far enough, not logging");
-                break;
+                tx, FlowGetDisruptionFlags(f, STREAM_TOCLIENT));
+            if (tx_progress == tx_progress_done_value_tc) {
+                tc_ready = 1;
             }
         }
 
+        if (!(ts_ready || tc_ready)) {
+            SCLogNotice("progress not for enough, not logging");
+            break;
+        }
+
+        /* Prevent the processing of one direction getting farther
```

Comment:
On 17-12-15 17:49, Jason Ish wrote:
> No, this patch is concerned with the more narrow scope of just logging
> the request before the response if available. This would preserve what I
> believe is the current behaviour or not logging until the response is
> received, correct?

The current behaviour is to log on 'tx complete', which means both
request and response are done.

If we move to a model where one callback is 'request complete' and the
other 'tx complete' (implying 'response complete') then I think the
request logging really should happen as soon as the request is complete.

Other transactions shouldn't affect this.

Perhaps the implied 'tx complete == response complete' doesn't always
make sense, and we need a separate 'response complete'. Not sure what
protocol would have a response before the request though :)